zebra-chain/src/orchard/shielded_data.rs

```diff
@@ -131,62 +136,94 @@ impl AtLeastOne<AuthorizedAction> {
+    pub const SPEND_AUTH_SIG_SIZE: u64 = 64;
+
+    /// The size of a single AuthorizedAction
+    ///
+    /// Each serialized `Action` has a corresponding `Signature<SpendAuth>`.
+    pub const AUTHORIZED_ACTION_SIZE: u64 = Self::ACTION_SIZE + Self::SPEND_AUTH_SIG_SIZE;
+
+    /// The maximum number of actions in the transaction.
+    // Since a serialized Vec<AuthorizedAction> uses at least one byte for its length,
+    // and the signature is required,
+    // a valid max allocation can never exceed this size
+    pub const ACTION_MAX_ALLOCATION: u64 = (MAX_BLOCK_BYTES - 1) / Self::AUTHORIZED_ACTION_SIZE;
+
     /// Split out the action and the signature for V5 transaction
     /// serialization.
-    pub fn into_parts(self) -> (Action, Signature<SpendAuth>) {
+    pub fn into_parts(self) -> (Action<V>, Signature<SpendAuth>) {
         (self.action, self.spend_auth_sig)
     }
 
     // Combine the action and the spend auth sig from V5 transaction
     /// deserialization.
-    pub fn from_parts(action: Action, spend_auth_sig: Signature<SpendAuth>) -> AuthorizedAction {
+    pub fn from_parts(
+        action: Action<V>,
+        spend_auth_sig: Signature<SpendAuth>,
+    ) -> AuthorizedAction<V> {
         AuthorizedAction {
             action,
             spend_auth_sig,
         }
     }
 }
 
-/// The size of a single Action
-///
-/// Actions are 5 * 32 + 580 + 80 bytes so the total size of each Action is 820 bytes.
-/// [7.5 Action Description Encoding and Consensus][ps]
-///
-/// [ps]: <https://zips.z.cash/protocol/nu5.pdf#actionencodingandconsensus>
-pub const ACTION_SIZE: u64 = 5 * 32 + 580 + 80;
-
-/// The size of a single `Signature<SpendAuth>`.
-///
-/// Each Signature is 64 bytes.
-/// [7.1 Transaction Encoding and Consensus][ps]
-///
-/// [ps]: <https://zips.z.cash/protocol/nu5.pdf#actionencodingandconsensus>
-pub const SPEND_AUTH_SIG_SIZE: u64 = 64;
+// TODO: FIXME: Consider moving it to transaction.rs as it's not used here. Or move its usage here from transaction.rs.
+/// A struct that contains references to several fields of an `Action` struct.
+/// Those fields are used in other parts of the code that call the `orchard_actions()` method of the `Transaction`.
+/// The goal of using `ActionRef` is that it's not a generic, unlike `Action`, so it can be returned from Transaction methods
+/// (the fields of `ActionRef` do not depend on the generic parameter `Version` of `Action`).
+pub struct ActionRef<'a> {
+    /// A reference to the value commitment to the net value of the input note minus the output note.
+    pub cv: &'a super::commitment::ValueCommitment,
+    /// A reference to the nullifier of the input note being spent.
+    pub nullifier: &'a super::note::Nullifier,
+    /// A reference to the randomized validating key for `spendAuthSig`.
+    pub rk: &'a reddsa::VerificationKeyBytes<SpendAuth>,
+    /// A reference to the x-coordinate of the note commitment for the output note.
+    pub cm_x: &'a pallas::Base,
+}
```

Comment:
Since this data is small, it would be ok to clone it. But if it works with lifetimes that's also ok.

For example, we clone anchors to abstract over v4 and v5 sapling transaction structures:
https://github.com/ZcashFoundation/zebra/blob/ca062d04d441f6e8011a8e1266861dbc5fd45721/zebra-chain/src/sapling/shielded_data.rs#L298-L322